### DISCUSSION TITLE
Unable to execute msprime.Demography() following msprime installation

### DISCUSSION THREAD

**Discussion by mshpak76** [tskit-dev/msprime#2068] [Q&A]
I followed the installation instructions for msprime and seemingly successfully completed the installation in Python 3.8.6 using pip, and I was able to execute:

     import msprime   
     testing=msprime.simulate(sample_size=5,Ne=1000000,mutation_rate=3e-7,recombination_rate=3e-7)

However, 

    demography = msprime.Demography()

gives the error 'module msprime has no attribute Demography(). The same error applies with msprime.sim_ancestry() and various other functions. Something seems to have gone wrong with the installation, but I don't know what it is and how to repair it.

Thank you in advance for any help in trouble-shooting this.

**Comment by benjeffery** [comment]
Hi @mshpak76!
I've transferred this issue to the msprime repo.
Can you tell us the result of `print(msprime.__version__)`?

**Reply by mshpak76** [reply]
0.7.4

**Reply by mshpak76** [reply]
I had also thought that perhaps it was due to working with an outdated version, but 0.7.4 is what came up by default with 

    pip install msprime

**Reply by benjeffery** [reply]
Interesting, that is quite an old version that doesn't have either `sim_ancestry` or `Demography`.
Can you try `pip install --no-cache-dir --upgrade msprime`

If that doesn't work can you:
```
pip uninstall msprime
pip install -vvv msprime
```
and share the output log.

**Reply by jeromekelleher** [reply]
It's probably a python-and-pip-on-different-versions-issue, I think.

Can you try using ``python -m pip install msprime`` rather than ``pip install msprime`` @mshpak76? (Where ``python`` is whatever the command line you use in order to start the python interpreter.)

**Reply by mshpak76** [reply]
I think I see where the problem is. I'm attempting a global installation, but for some reason sudo pip install msprime is installing msprime version 1.2.0 in /usr/local/bin rather than the opt/bin where Python 3.8.6 is globally installed. So when I run Python, the old version of msprime 0.7.4 remains (similarly, if I try to use sudo pip uninstall msprime, it will attempt to uninstall msprime 1.2.0 in my local directory). So the problem is with what sudo pip is doing rather than any issue with msprime, though I have no idea why it is doing this.

**Reply by benjeffery** [reply]
I _strongly_ recommend using a virtual environment. See [here](https://realpython.com/python-virtual-environments-a-primer/) for a guide.

**Reply by jeromekelleher** [reply]
sudo python -m pip should do what you want. It's almost never a good idea to use the pip executable directly.

**Reply by mshpak76** [reply]
Thanks - I have msprime 1.2 installed both globally and in a conda environment (associated with a newer version of python) now.
